File: data_structures_py/mydeque.py

class MyDeque():
	def __init__(self, max=5):
		self.front = -1
		self.rear = -1
		self.arr = [0]*max
		self.max = max

	def isEmpty(self):
		return (self.front == -1);

	def isFull(self):
		return ( (self.rear-self.front+self.max)%self.max == self.max-1 );

	def insertFront(self, item):
		if (self.isFull()):
			print("OverFlow at Insertion")
		
		if (self.isEmpty()):
			self.front = 1;
			self.rear = 0;
		
		self.front = (self.front - 1 + self.max) % self.max;   # keeping it circular
		self.arr[self.front] = item;
	
	def insertRear(self, item):
		if (self.isFull()):
			print("OverFlow at Insertion")
		
		if (self.isEmpty()):
			self.front = 0;
			self.rear = -1;
		
		self.rear = (self.rear + 1 + self.max) % self.max;   # keeping it circular
		self.arr[self.rear] = item;
	
	def deleteFront(self):
		if (self.isEmpty()):
			print("Cannot remove from Empty Queue")
			return
		
		if (self.front == self.rear):
			self.front = -1;
			self.rear = -1;
		else:
			self.front = (self.front + 1) % self.max;
		
	def deleteRear(self):
		if (self.isEmpty()):
			print("Cannot remove from Empty Queue")
			return
		
		if (self.front == self.rear):
			self.front = -1;
			self.rear = -1;
		else:
			self.rear = (self.rear - 1 + self.max) % self.max;
		
	def printQueue(self):
		if (self.isEmpty()):
			print("Empty Queue")
			return
		
		iters = ( self.rear-self.front+self.max ) % self.max;
		for i in range(iters+1):
			print( self.arr[ (self.front+i)%self.max ] )
		
	


if __name__ == "__main__":
	d = MyDeque()

	d.insertFront(34);
	d.insertFront(12);
	d.insertRear(120);
	d.deleteFront();
	d.deleteRear();

	d.printQueue();

File: data_structures_js/mydeque.js
class MyDeque {
	constructor (max = 5) {
		this.max = max;
		this.front = -1;
		this.rear = -1;
		this.arr = [];
	}
	isEmpty() {
		return this.front == -1;
	}
	isFull() {
		return ( (this.rear-this.front+this.max)%this.max == this.max-1 );
	}
	insertFront(item) {
		if (this.isFull()) {
			console.log("OverFlow at Insertion");
			return;
		}
		if (this.isEmpty()) {
			this.front = 1;
			this.rear = 0;
		}
		this.front = (this.front - 1 + this.max) % this.max;   // keeping it circular
		this.arr[this.front] = item;

	}
	insertRear(item) {
		if (this.isFull()) {
			console.log("OverFlow at Insertion");
			return;
		}
		if (this.isEmpty()) {
			this.front = 0;
			this.rear = -1;
		}
		this.rear = (this.rear + 1 + this.max) % this.max;   // keeping it circular
		this.arr[this.rear] = item;
	}
	deleteFront() {
		if (this.isEmpty()) {
			console.log("Cannot remove from Empty Queue");
		}
		if (this.front == this.rear) {
			this.front = -1;
			this.rear = -1;
		} else {
			this.front = (this.front + 1) % this.max;
		}
	}
	deleteRear() {
		if (this.isEmpty()) {
			console.log("Cannot remove from Empty Queue");
		}
		if (this.front == this.rear) {
			this.front = -1;
			this.rear = -1;
		} else {
			this.rear = (this.rear - 1 + this.max) % this.max;
		}
	}
	printQueue() {
		if (this.isEmpty()) {
			console.log("Empty Queue"); return;
		}
		let iters = ( this.rear-this.front+this.max ) % this.max;
		for (let i=0; i<=iters; i++) {
			console.log( this.arr[ (this.front+i)%this.max ] )
		}
	}
}

if (require.main == module) {
	d = new MyDeque();

	d.insertFront(34);
	d.insertFront(12);
	d.insertRear(120);
	d.deleteFront();
	d.deleteRear();

	d.printQueue();

}